plugins/ocra/hu.bme.mit.gamma.ocra.transformation/src/hu/bme/mit/gamma/ocra/transformation/ModelSerializer.xtend

```diff
@@ -23,17 +36,174 @@ class ModelSerializer {
 		
 		return component.execute
 	}
 	
-	def String execute(Component component) '''
-		COMPONENT «component.name» system
-	'''
+	def String execute(Component component) {
+        '''
+        «serializeSystemComponent(component)»
+        «serializeSubComponents(getSubComponents(component))»
+        '''
+	}
+	
+      def String serializeSystemComponent(Component component) {
+      	val ports = component.allPorts
+        '''COMPONENT «component.name» system
+        		   INTERFACE
+        			«serializeInterface(ports)»
+        		   REFINEMENT
+        			«serializeRefinement(component)»
+        		'''
+    }
+    
+    
+	def String serializeSubComponents(Set<Component> components) {	
+		'''«FOR component : components»
+			COMPONENT «component.name»
+					INTERFACE
+			        			«serializeInterface(component.ports)»
+					REFINEMENT
+			        			«serializeRefinement(component)»
+			        			
+		«ENDFOR»'''
+	}
+
+	def String serializeInterface(List<Port> ports) {
+        '''
+        «FOR port : ports»
+	        «FOR event : port.inputEvents»
+	        INPUT «naming.getPortName(port, event)»
+	        «ENDFOR»
+	        «FOR event : port.outputEvents»
+	        OUTPUT «naming.getPortName(port, event)»
+	        «ENDFOR»
+        «ENDFOR»
+        
+        ''' 
+    }
+    
+	def serializeRefinement(Component component) {
+		val subcomponents = extractSubcomponentInstances(component)
+		val bindings = extractBindings(component)
+		val channels = extractChannels(component)
+		if(!subcomponents.nullOrEmpty) {
+		'''
+		«FOR sub: subcomponents»
+		«naming.getSubName(sub)»
+		«ENDFOR»
+		
+		«FOR binding : bindings»
+			«serializeBinding(binding)»
+		«ENDFOR»
+		«FOR channel : channels»
+			«serializeChannel(channel)»
+		«ENDFOR»
+		
+		'''
+		
+		}
+	}
+	
+		
+	def String serializeChannel(Channel channel) {
+		val events = channel.providedPort.port.allEvents
+		val ports = channel.requiredPorts
+		'''
+		«FOR port : ports»
+			«FOR event : events»
+			«naming.getChannelName(channel, port, event)»
+			«ENDFOR»
+		«ENDFOR»
+		'''
+	}
+	
+	def serializeBinding(PortBinding binding) {
+		'''
+		«FOR event : binding.compositeSystemPort.allEvents»
+			«naming.getBindingName(binding, event)»
+		«ENDFOR»
+		'''
+
+	}
+		
+	def List<ComponentInstance> extractSubcomponentInstances(Component component) {
+		val derivedComponents = new ArrayList
+		if(component instanceof CompositeComponent) {
+			for(instance : component.derivedComponents) {
+				derivedComponents.add(instance)
+			}
+			return derivedComponents
+		}
+	}
 	
-}
\ No newline at end of file
+
+	def EList<PortBinding> extractBindings(Component component) {
+		if(component instanceof CompositeComponent) {
+			return component.portBindings			
+		}
+	}
+	
+	def List<Channel> extractChannels(Component component) {
+	if(component instanceof CompositeComponent) {
+			return component.channels			
+		}	
+	}
+		
+	def Set<Component> getSubComponents(Component component) {
```

Comment:
I believe there is a method in `StatechartModelDerivedFeatures` with the same functionality. Use that and delete this

Reply:
Since the uniqueness of the component is required i could not find other solution than mapping it into a set